### PR TITLE
Event observable and strongly typed response objects

### PR DESCRIPTION
**Pull request by tramonex-nate**
Made a few improvements to how the contract wrappers represent event.  I think it's a step in the right direction, but I'm open to suggestions on how to improve this more!

### CODE REVIEW

**Comment by conor10** [comment]
Thanks for submitting this.

Would you be able to update the generated [HumanStandardToken](https://github.com/web3j/web3j/blob/master/src/integration-test/java/org/web3j/generated/HumanStandardToken.java) class file with this modified version of the generator, and the associated [HumanStandardTokenGeneratedIT](https://github.com/web3j/web3j/blob/master/src/integration-test/java/org/web3j/protocol/scenarios/HumanStandardTokenGeneratedIT.java)?

This will help demonstrate the changes.

I should get around to properly reviewing the changes next week.

File: src/main/java/org/web3j/codegen/SolidityFunctionWrapperGenerator.java [hunk 1]
```diff
@@ -432,37 +439,99 @@ private static void buildTransactionFunction(
...
-        String generatedFunctionName =
-                "process" + Strings.capitaliseFirstLetter(functionName) + "Event";
+        for (NamedTypeName namedType : nonIndexedParameters) {
+            builder.addField(namedType.getTypeName(), namedType.getName(), Modifier.PUBLIC);
+        }
+
+        return builder.build();
+    }
+
+    static MethodSpec buildEventObservableFunction(String responseClassName, String functionName, List<NamedTypeName> indexedParameters, List<NamedTypeName> nonIndexedParameters) throws ClassNotFoundException {
```
Comment (conor10) [inline src/main/java/org/web3j/codegen/SolidityFunctionWrapperGenerator.java]: Please can you ensure you keep lines to less than 100 chars in length. I haven't added checkstyle constraints just yet.

File: src/main/java/org/web3j/codegen/SolidityFunctionWrapperGenerator.java [hunk 2]
```diff
@@ -432,37 +439,99 @@ private static void buildTransactionFunction(
...
+
+        return observableMethodBuilder
+                .build();
+    }
 
-        MethodSpec.Builder methodBuilder = MethodSpec.methodBuilder(generatedFunctionName)
+    static MethodSpec buildEventTransactionReceiptFunction(String responseClassName, String functionName, List<NamedTypeName> indexedParameters, List<NamedTypeName> nonIndexedParameters) throws ClassNotFoundException {
+        ParameterizedTypeName parameterizedTypeName = ParameterizedTypeName.get(ClassName.get(List.class), ClassName.get("", responseClassName));
+
+        String generatedFunctionName = "get" + Strings.capitaliseFirstLetter(functionName) + "EventFromTransactionReceipt";
```
Comment (conor10) [inline src/main/java/org/web3j/codegen/SolidityFunctionWrapperGenerator.java]: You don't need to include "FromTransactionReceipt" in the generated method name, as the TransactionReceipt is an explicit parameter.

File: src/main/java/org/web3j/codegen/SolidityFunctionWrapperGenerator.java [hunk 3]
```diff
@@ -432,37 +439,99 @@ private static void buildTransactionFunction(
...
+                .addModifiers(Modifier.PUBLIC)
+                .returns(parameterizedTypeName);
+
+
+        buildVariableLengthEventConstructor(
+                observableMethodBuilder, functionName, indexedParameters, nonIndexedParameters);
+
+        observableMethodBuilder.addStatement("$1T filter = new $1T($2T.EARLIEST,$2T.LATEST, getContractAddress())", EthFilter.class, DefaultBlockParameterName.class)
+                .addStatement("filter.addSingleTopic($T.encode(event))", EventEncoder.class)
+                .beginControlFlow("return web3j.ethLogObservable(filter).map(log ->")
```
Comment (conor10) [inline src/main/java/org/web3j/codegen/SolidityFunctionWrapperGenerator.java]: Please can you use Java 6 friendly syntax for the generated code? This way the generated code remains web3j-android compatible.

**Comment by conor10** [comment]
BTW - the TransactionManager changes I've made in web3j should be straight forwards to merge.

**Comment by tramonex-nate** [comment]
Cool, I'll should be able to get around to making those changes this or next week, and updating the generated HumanStandardToken!

**Comment by conor10** [comment]
Great - I'll hold off putting out the 1.2.0 web3j release until you've submitted this change.

Please let me know if this gets delayed on your side.

**Comment by tramonex-nate** [comment]
@conor10, just got around to addressing all the changes, have a look and see if that works for you!

**Comment by conor10** [comment]
Thanks a lot for submitting this PR - it's a great enhancement to web3j!

**Pull request merged**